File: Online-judge/leetcode/minimum-distance-between-bst-nodes.cpp
# include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/maximum-depth-of-binary-tree/description/
 * Running complexity O(n)
 * Memory complexity O(log(n))
*/


class Solution {
private:
    vector<int> nodes;

    void dfs(TreeNode* root) {
        if (!root) return;
        nodes.push_back(root->val);
        dfs(root->left); dfs(root->right);
    }
    
public:
    int minDiffInBST(TreeNode* root) {
        dfs(root);
        sort(nodes.begin(), nodes.end());
        int res = INT_MAX;
        
        for (int i = 1; i < nodes.size(); i++) {
            if (nodes[i] - nodes[i-1] < res) {
                res = nodes[i] - nodes[i-1];
            }
        }
        return res;
    }
};

File: Online-judge/leetcode/minimum-distance-between-bst-nodes.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    nodes: List = []
    def dfs(self, root: Optional[TreeNode]):
        if (root == None):
            return
        self.nodes.append(root.val)
        self.dfs(root.left)
        self.dfs(root.right)

    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.nodes.clear()
        self.dfs(root)
        self.nodes.sort()
        res: int = 10000000

        for index in range(1, len(self.nodes)):
            if (self.nodes[index] - self.nodes[index-1] < res):
                res = self.nodes[index] - self.nodes[index-1]
        return res